Update numpy hook for compatibility with 1.22

rokm opened the pull request:
As of `numpy` 1.22.0, `numpy.testing` [imports `numpy.testing._private.extbuild`](https://github.com/numpy/numpy/blob/4adc87dff15a247e417d50f10cc4def8e1c17a03/numpy/testing/__init__.py#L12), which in turn imports `numpy.distutils` and `distutils` that we explicitly [exclude in the hook](https://github.com/pyinstaller/pyinstaller/blob/db95de4fa6c88a96d04cfb7a94e1d06b3bb3813a/PyInstaller/hooks/hook-numpy.py#L44-L53). 

The `numpy.testing` is imported by `scipy.optimize` (via `scipy.optimize._constraints`); therefore, a frozen application using that module and numpy 1.22 ends up with missing `numpy.distutils` module error, as seen in #6473.

But before we tackle `numpy` hook itself, the scipy tests need to be cleaned up a bit - the lack of failure on Windows and macOS runners due to xfail does not exactly inspire confidence...

rokm committed:
tests: remove xfail annotations from scipy tests

The referenced issues do not seem to be actual anymore. On the
other hand, the xfail annotation hides the fact that scipy tests
end up broken with numpy 1.22.

rokm committed:
tests: scipy: remove obsolete version check

We are unlikely to encounter scipy < 0.16 under python versions
that we support. The removed version comparison, while harmless
by itself, used distutils.LooseVersion, so removing the check
also allows us to remove dependency on distutils for this
particular test.

rokm committed:
hooks: numpy: do not exclude distutils and numpy.distutils anymore

As of numpy 1.22, distutils and numpy.distutils are required by
numpy.testing (due to added import of numpy.testing._private.extbuild),
which in turn is used for example by scipy.optimize (via its
scipy.optimize._constraints module). So we cannot exclude them
anymore, except for earlier numpy versions.

bwoodsend commented:
[Sigh], yet again I find myself wishing that NumPy's test suite and building machinery where not part of the NumPy package itself.

rokm commented:
Yeah, the best part is that `scipy` is [importing](https://github.com/scipy/scipy/blob/a0e129fc3ba89fd81a5fa7e5e4d6aa2ec635a674/scipy/optimize/_constraints.py#L8) `numpy.testing` for its `suppres_warnings` context_manager/decorator. Based on the [docs for that one](https://github.com/numpy/numpy/blob/4adc87dff15a247e417d50f10cc4def8e1c17a03/numpy/testing/_private/utils.py#L2032-L2042), I wonder if they could simply go with `warnings.catch_warnings` instead...

rokm committed:
tests: repin numpy version

For python 3.7, use 1.21.5, which is the last release supporting
python 3.7.

Use 1.22 with python 3.8 and newer, including python 3.10.

bwoodsend commented:
> Yeah, the best part is that `scipy` is [importing](https://github.com/scipy/scipy/blob/a0e129fc3ba89fd81a5fa7e5e4d6aa2ec635a674/scipy/optimize/_constraints.py#L8) `numpy.testing` for its `suppres_warnings` context_manager/decorator. Based on the [docs for that one](https://github.com/numpy/numpy/blob/4adc87dff15a247e417d50f10cc4def8e1c17a03/numpy/testing/_private/utils.py#L2032-L2042), I wonder if they could simply go with `warnings.catch_warnings` instead...

I'll try it as a pull request and see what they say. At the very least, having that dependency knocks up the `import scipy` initialisation time.

bwoodsend commented:
Although, now that I remember that scipy's test suite takes about a day and a half to run, this may take a while.

rokm commented:
This takes care of of `numpy`. 

I noticed `scipy` 1.8.0 release candidates have also dropped python 3.7 support, and other projects following NEP 29 will likely follow the suit. So perhaps we should pre-emptively pin the python 3.7 dependencies for projects where we already have 3.6-specific pins (e.g., `numpy`, `scipy`, `matplotlib`) to the current version... (And we might also consider retiring 3.6 runners so that all macOS runners take only one cycle).

bwoodsend commented:
I imagine that `matplotlib` and `pandas` will follow suit but I think I'd rather know than assume that they will. Python 3.6 officially died 8 days ago so we might as well reclaim the 3.6 runners and drop 3.6 support.

bwoodsend commented:
> Yeah, the best part is that `scipy` is [importing](https://github.com/scipy/scipy/blob/a0e129fc3ba89fd81a5fa7e5e4d6aa2ec635a674/scipy/optimize/_constraints.py#L8) `numpy.testing` for its `suppres_warnings` context_manager/decorator. Based on the [docs for that one](https://github.com/numpy/numpy/blob/4adc87dff15a247e417d50f10cc4def8e1c17a03/numpy/testing/_private/utils.py#L2032-L2042), I wonder if they could simply go with `warnings.catch_warnings` instead...

There are a couple of others which are only loaded by some submodules. But it's the same story for each of them - they're all trivially replaceable.

```
/e/n/scipy (dont-import-numpy-distutils)> git grep numpy.testing scipy | grep --invert-match /tests
scipy/special/_mptestutils.py:from numpy.testing import assert_
scipy/special/_test_round.pyx:from numpy.testing import assert_
scipy/special/_testutils.py:from numpy.testing import assert_
scipy/stats/_binned_statistic.py:from numpy.testing import suppress_warnings
```
Although only the last is actually loaded on importing anything public.
